File: app/src/main/java/com/dpis/module/ModulePackagePlan.java
package com.dpis.module;

import com.dpis.module.runtime.appprocess.AppProcessHookInstaller;

import com.dpis.module.viewport.ViewportApplyMode;
import com.dpis.module.viewport.ViewportTargetSpec;

import com.dpis.module.hooks.HookDomainOverride;
import com.dpis.module.hooks.HookExecutionPlan;
import com.dpis.module.hooks.HookExecutionPlanner;
import com.dpis.module.hooks.HookRuntimePolicy;

public final class ModulePackagePlan {
    public final String packageName;
    public final ViewportTargetSpec targetViewportSpec;
    public final String targetViewportMode;
    public final Integer targetFontScalePercent;
    public final String targetFontMode;
    public final String targetTypefaceId;
    public final boolean targetDpisEnabled;
    public final boolean viewportConfigured;
    public final boolean viewportEnabled;
    public final boolean fontScaleActive;
    public final boolean fontEnabled;
    public final boolean typefaceActive;
    public final boolean typefaceEnabled;
    public final boolean flutterSettingsFontEnabled;
    public final boolean hyperOsNativeFlutterFontEnabled;
    public final HookDomainOverride hookDomainOverride;

    private ModulePackagePlan(String packageName,
                              ViewportTargetSpec targetViewportSpec,
                              String targetViewportMode,
                              Integer targetFontScalePercent,
                              String targetFontMode,
                              String targetTypefaceId,
                              boolean targetDpisEnabled,
                              boolean viewportConfigured,
                              boolean viewportEnabled,
                              boolean fontScaleActive,
                              boolean fontEnabled,
                              boolean typefaceActive,
                              boolean typefaceEnabled,
                              boolean flutterSettingsFontEnabled,
                              boolean hyperOsNativeFlutterFontEnabled,
                              HookDomainOverride hookDomainOverride) {
        this.packageName = packageName;
        this.targetViewportSpec = targetViewportSpec != null
                ? targetViewportSpec
                : ViewportTargetSpec.off();
        this.targetViewportMode = targetViewportMode;
        this.targetFontScalePercent = targetFontScalePercent;
        this.targetFontMode = targetFontMode;
        this.targetTypefaceId = targetTypefaceId;
        this.targetDpisEnabled = targetDpisEnabled;
        this.viewportConfigured = viewportConfigured;
        this.viewportEnabled = viewportEnabled;
        this.fontScaleActive = fontScaleActive;
        this.fontEnabled = fontEnabled;
        this.typefaceActive = typefaceActive;
        this.typefaceEnabled = typefaceEnabled;
        this.flutterSettingsFontEnabled = flutterSettingsFontEnabled;
        this.hyperOsNativeFlutterFontEnabled = hyperOsNativeFlutterFontEnabled;
        this.hookDomainOverride = hookDomainOverride != null
                ? hookDomainOverride
                : HookDomainOverride.automatic();
    }

    static ModulePackagePlan resolve(DpisConfigStore store, String packageName) {
        return resolve(ConfigSnapshotLoader.fromStore(store), packageName);
    }

    static ModulePackagePlan resolve(ConfigSnapshot snapshot, String packageName) {
        if (snapshot == null || packageName == null || packageName.isBlank()
                || !snapshot.isConfigured(packageName)) {
            return inactive(packageName);
        }
        PackageConfigSnapshot packageConfig = snapshot.getPackage(packageName);
        if (packageConfig == null) {
            return inactive(packageName);
        }
        ViewportTargetSpec targetViewportSpec = packageConfig.targetViewportSpec;
        String targetViewportMode = packageConfig.targetViewportMode;
        Integer targetFontScalePercent = packageConfig.targetFontScalePercent;
        String targetFontMode = packageConfig.targetFontMode;
        String targetTypefaceId = packageConfig.targetTypefaceId;
        boolean targetDpisEnabled = packageConfig.dpisEnabled;
        boolean hyperOsNativeFlutterFontEnabled = packageConfig.flutterFontHookEnabled
                && packageConfig.hyperOsFlutterFontHookEnabled;
        boolean flutterSettingsFontEnabled = packageConfig.flutterFontHookEnabled
                && packageConfig.flutterSettingsFontHookEnabled;
        boolean fontScaleActive = targetFontScalePercent != null
                && targetFontScalePercent > 0
                && targetFontScalePercent != 100;
        boolean typefaceActive = targetTypefaceId != null && !targetTypefaceId.isBlank();
        if (!targetDpisEnabled
                || (!targetViewportSpec.isEnabled() && !fontScaleActive && !typefaceActive)) {
            return new ModulePackagePlan(
                    packageName,
                    targetViewportSpec,
                    targetViewportMode,
                    targetFontScalePercent,
                    targetFontMode,
                    targetTypefaceId,
                    targetDpisEnabled,
                    targetViewportSpec.isEnabled(),
                    false,
                    fontScaleActive,
                    false,
                    typefaceActive,
                    false,
                    flutterSettingsFontEnabled,
                    hyperOsNativeFlutterFontEnabled,
                    packageConfig.hookDomainOverride);
        }
        HookRuntimePolicy policy = HookRuntimePolicy.fromSnapshot(snapshot);
        boolean viewportConfigured = targetViewportSpec.isEnabled();
        boolean viewportEnabled = AppProcessHookInstaller.resolveViewportHookEnabled(
                policy, viewportConfigured, targetViewportMode);
        AppProcessHookInstaller.FontHookPlan fontHookPlan =
                AppProcessHookInstaller.resolveFontHookPlan(
                        policy, fontScaleActive, targetFontMode);
        return new ModulePackagePlan(
                packageName,
                targetViewportSpec,
                targetViewportMode,
                targetFontScalePercent,
                targetFontMode,
                targetTypefaceId,
                targetDpisEnabled,
                viewportConfigured,
                viewportEnabled,
                fontScaleActive,
                fontHookPlan.emulationEnabled || fontHookPlan.fieldRewriteEnabled,
                typefaceActive,
                typefaceActive,
                flutterSettingsFontEnabled,
                hyperOsNativeFlutterFontEnabled,
                packageConfig.hookDomainOverride);
    }

    boolean shouldInstallHooks() {
        // Flutter supplements are currently attached only when a normal font route is active.
        // If they become standalone routes, include them in this predicate as well.
        return targetDpisEnabled && (viewportEnabled || fontEnabled || typefaceEnabled);
    }

    boolean shouldInstallLegacyHooks() {
        return targetDpisEnabled
                && (viewportEnabled
                || typefaceEnabled
                || (fontScaleActive
                && FontApplyMode.isEnabled(targetFontMode)));
    }

    boolean hasSecondaryProcessSafeRoute() {
        return fontEnabled || typefaceEnabled;
    }

    ModulePackagePlan withoutViewportRoute() {
        return new ModulePackagePlan(
                packageName,
                ViewportTargetSpec.off(),
                ViewportApplyMode.OFF,
                targetFontScalePercent,
                targetFontMode,
                targetTypefaceId,
                targetDpisEnabled,
                false,
                false,
                fontScaleActive,
                fontEnabled,
                typefaceActive,
                typefaceEnabled,
                flutterSettingsFontEnabled,
                hyperOsNativeFlutterFontEnabled,
                hookDomainOverride);
    }

    public HookExecutionPlan buildExecutionPlan(HookRuntimePolicy policy, DebugFontOverride debugOverride) {
        return HookExecutionPlanner.buildPlan(
                policy,
                packageName,
                viewportConfigured,
                targetViewportMode,
                fontScaleActive,
                targetFontMode,
                flutterSettingsFontEnabled,
                hyperOsNativeFlutterFontEnabled,
                hookDomainOverride,
                debugOverride);
    }

    Integer targetViewportWidthDp() {
        return targetViewportSpec.isAbsoluteDp()
                ? targetViewportSpec.absoluteWidthDp()
                : null;
    }

    private static ModulePackagePlan inactive(String packageName) {
        return new ModulePackagePlan(
                packageName,
                ViewportTargetSpec.off(),
                ViewportApplyMode.OFF,
                null,
                FontApplyMode.OFF,
                null,
                false,
                false,
                false,
                false,
                false,
                false,
                false,
                false,
                false,
                HookDomainOverride.automatic());
    }
}

File: app/src/main/java/com/dpis/module/config/ModulePackagePlan.kt
package com.dpis.module.config

import com.dpis.module.fonts.FontApplyMode
import com.dpis.module.hooks.HookDomainOverride
import com.dpis.module.hooks.HookExecutionPlan
import com.dpis.module.hooks.HookExecutionPlanner
import com.dpis.module.hooks.HookRuntimePolicy
import com.dpis.module.runtime.appprocess.AppProcessHookInstaller
import com.dpis.module.runtime.font.DebugFontOverride
import com.dpis.module.viewport.ViewportApplyMode
import com.dpis.module.viewport.ViewportTargetSpec

class ModulePackagePlan private constructor(
    @JvmField val packageName: String?,
    targetViewportSpec: ViewportTargetSpec?,
    @JvmField val targetViewportMode: String?,
    @JvmField val targetFontScalePercent: Int?,
    @JvmField val targetFontMode: String?,
    @JvmField val targetTypefaceId: String?,
    @JvmField val targetDpisEnabled: Boolean,
    @JvmField val viewportConfigured: Boolean,
    @JvmField val viewportEnabled: Boolean,
    @JvmField val fontScaleActive: Boolean,
    @JvmField val fontEnabled: Boolean,
    @JvmField val typefaceActive: Boolean,
    @JvmField val typefaceEnabled: Boolean,
    @JvmField val flutterSettingsFontEnabled: Boolean,
    @JvmField val hyperOsNativeFlutterFontEnabled: Boolean,
    hookDomainOverride: HookDomainOverride?,
) {
    @JvmField val targetViewportSpec = targetViewportSpec ?: ViewportTargetSpec.off()
    @JvmField val hookDomainOverride = hookDomainOverride ?: HookDomainOverride.automatic()

    fun shouldInstallHooks(): Boolean =
        targetDpisEnabled && (viewportEnabled || fontEnabled || typefaceEnabled)

    fun shouldInstallLegacyHooks(): Boolean =
        targetDpisEnabled && (
            viewportEnabled ||
                typefaceEnabled ||
                (fontScaleActive && FontApplyMode.isEnabled(targetFontMode))
            )

    fun hasSecondaryProcessSafeRoute(): Boolean = fontEnabled || typefaceEnabled

    fun withoutViewportRoute(): ModulePackagePlan = create(
        packageName,
        ViewportTargetSpec.off(),
        ViewportApplyMode.OFF,
        targetFontScalePercent,
        targetFontMode,
        targetTypefaceId,
        targetDpisEnabled,
        false,
        false,
        fontScaleActive,
        fontEnabled,
        typefaceActive,
        typefaceEnabled,
        flutterSettingsFontEnabled,
        hyperOsNativeFlutterFontEnabled,
        hookDomainOverride,
    )

    fun buildExecutionPlan(
        policy: HookRuntimePolicy,
        debugOverride: DebugFontOverride?,
    ): HookExecutionPlan = HookExecutionPlanner.buildPlan(
        policy,
        packageName,
        viewportConfigured,
        targetViewportMode,
        fontScaleActive,
        targetFontMode,
        flutterSettingsFontEnabled,
        hyperOsNativeFlutterFontEnabled,
        hookDomainOverride,
        debugOverride,
    )

    fun targetViewportWidthDp(): Int? =
        if (targetViewportSpec.isAbsoluteDp()) targetViewportSpec.absoluteWidthDp() else null

    companion object {
        @JvmStatic
        fun resolve(store: ConfigSnapshotStore?, packageName: String?): ModulePackagePlan =
            resolve(ConfigSnapshotLoader.fromStore(store), packageName)

        @JvmStatic
        fun resolve(snapshot: ConfigSnapshot?, packageName: String?): ModulePackagePlan {
            if (snapshot == null || packageName.isNullOrBlank() || !snapshot.isConfigured(packageName)) {
                return inactive(packageName)
            }
            val config = snapshot.getPackage(packageName) ?: return inactive(packageName)
            val viewport = config.targetViewportSpec
            val fontScale = config.targetFontScalePercent
            val typeface = config.targetTypefaceId
            val fontScaleActive = fontScale != null && fontScale > 0 && fontScale != 100
            val typefaceActive = !typeface.isNullOrBlank()
            val flutterSettings = config.flutterFontHookEnabled && config.flutterSettingsFontHookEnabled
            val hyperOs = config.flutterFontHookEnabled && config.hyperOsFlutterFontHookEnabled
            if (!config.dpisEnabled || (!viewport.isEnabled() && !fontScaleActive && !typefaceActive)) {
                return create(
                    packageName,
                    viewport,
                    config.targetViewportMode,
                    fontScale,
                    config.targetFontMode,
                    typeface,
                    config.dpisEnabled,
                    viewport.isEnabled(),
                    false,
                    fontScaleActive,
                    false,
                    typefaceActive,
                    false,
                    flutterSettings,
                    hyperOs,
                    config.hookDomainOverride,
                )
            }
            val policy = HookRuntimePolicy.fromSnapshot(snapshot)
            val viewportEnabled = AppProcessHookInstaller.resolveViewportHookEnabled(
                policy, viewport.isEnabled(), config.targetViewportMode,
            )
            val fontPlan = AppProcessHookInstaller.resolveFontHookPlan(
                policy, fontScaleActive, config.targetFontMode,
            )
            return create(
                packageName,
                viewport,
                config.targetViewportMode,
                fontScale,
                config.targetFontMode,
                typeface,
                config.dpisEnabled,
                viewport.isEnabled(),
                viewportEnabled,
                fontScaleActive,
                fontPlan.emulationEnabled || fontPlan.fieldRewriteEnabled,
                typefaceActive,
                typefaceActive,
                flutterSettings,
                hyperOs,
                config.hookDomainOverride,
            )
        }

        private fun inactive(packageName: String?) = create(
            packageName,
            ViewportTargetSpec.off(),
            ViewportApplyMode.OFF,
            null,
            FontApplyMode.OFF,
            null,
            false,
            false,
            false,
            false,
            false,
            false,
            false,
            false,
            false,
            HookDomainOverride.automatic(),
        )

        private fun create(
            packageName: String?,
            viewport: ViewportTargetSpec?,
            viewportMode: String?,
            fontScale: Int?,
            fontMode: String?,
            typeface: String?,
            dpisEnabled: Boolean,
            viewportConfigured: Boolean,
            viewportEnabled: Boolean,
            fontScaleActive: Boolean,
            fontEnabled: Boolean,
            typefaceActive: Boolean,
            typefaceEnabled: Boolean,
            flutterSettings: Boolean,
            hyperOs: Boolean,
            override: HookDomainOverride?,
        ) = ModulePackagePlan(
            packageName,
            viewport,
            viewportMode,
            fontScale,
            fontMode,
            typeface,
            dpisEnabled,
            viewportConfigured,
            viewportEnabled,
            fontScaleActive,
            fontEnabled,
            typefaceActive,
            typefaceEnabled,
            flutterSettings,
            hyperOs,
            override,
        )
    }
}
